src/theory/fp/theory_fp_rewriter.cpp

```diff
@@ -12,17 +12,22 @@
  **
  ** \brief [[ Rewrite rules for floating point theories. ]]
  **
- ** \todo [[ Constant folding
- **          Push negations up through arithmetic operators (include max and min? maybe not due to +0/-0)
+ ** \todo [[ Single argument constant propagate / simplify
+             Push negations through arithmetic operators (include max and min? maybe not due to +0/-0)
  **          classifications to normal tests (maybe)
  **          (= x (fp.neg x)) --> (isNaN x)
  **          (fp.eq x (fp.neg x)) --> (isZero x)   (previous and reorganise should be sufficient)
- **          (fp.eq x const) --> various = depending on const 
+ **          (fp.eq x const) --> various = depending on const
  **          (fp.abs (fp.neg x)) --> (fp.abs x)
  **          (fp.isPositive (fp.neg x)) --> (fp.isNegative x)
  **          (fp.isNegative (fp.neg x)) --> (fp.isPositive x)
  **          (fp.isPositive (fp.abs x)) --> (not (isNaN x))
  **          (fp.isNegative (fp.abs x)) --> false
+ **          A -> castA --> A
+ **          A -> castB -> castC  -->  A -> castC if A <= B <= C
+ **          A -> castB -> castA  -->  A if A <= B
+ **          promotion converts can ignore rounding mode
+ **          Samuel Figuer results
```

Comment:
Not quite sure how to parse this. Would you mind expanding on this (can be on top of all other changes if that's easier).

Reply:
Ah; sorry, these were some of my TODO notes.  It's a reference to a paper by Samuel Figuer on when double rounding is not a problem.  Basically it says:

float add (float f, float g) {
  return (float)( (double)f + (double)g );
}

is the same as a float add.  This is *really* useful to work around inadvertant type promotion in C.

How about I just implement it at some point?

Reply:
Maybe we could open an issue for that s.t. we don't forget about it?

Reply:
I opened an issue: #1149. I couldn't assign you @martin-cs , so I assigned myself because I know what this is about.